src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                     builder.ElementTypeWithAnnotations = collectionType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Single();
 
                     MethodSymbol getEnumeratorMethod = GetGetEnumeratorMethod(collectionType, diagnostics, isAsync, errorLocationSyntax);
+                    MethodSymbol moveNextMethod = null;
 
                     if ((object)getEnumeratorMethod != null)
                     {
                         MethodSymbol specificGetEnumeratorMethod = getEnumeratorMethod.AsMember(collectionType);
-                        builder.GetEnumeratorMethod = specificGetEnumeratorMethod;
                         TypeSymbol enumeratorType = specificGetEnumeratorMethod.ReturnType;
 
                         if (isAsync)
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
+
+                            // IAsyncEnumerable<T>.GetAsyncEnumerator has a default param, so let's fill it in
```

Comment:
>// IAsyncEnumerable<T>.GetAsyncEnumerator has a default param, so let's fill it in [](start = 28, length = 82)

How do we know that? How do we know we are dealing with ```IAsyncEnumerable<T>.GetAsyncEnumerator```? How do we know that there is a default parameter to fill in, even if it is a ```IAsyncEnumerable<T>.GetAsyncEnumerator```? Where did we check that? We cannot make an assumption like that solely based on the fact that this is how the method is defined in the standard framework.

Reply:
Extracted into a separate method to make it clear we're dealing with WellKnown or Special members.

---
In reply to: [549815596](https://github.com/dotnet/roslyn/pull/50105#discussion_r549815596) [](ancestors = 549815596)